lib/excalt/vcard/contact.ex

```diff
@@ -1,83 +1,196 @@
 defmodule Excalt.Vcard.Contact do
   @moduledoc """
-  Single Vcard contact
+  Module responsible for CRUD operations on Vcards
   """
 
+  @type t :: %__MODULE__{
+          etag: String.t(),
+          url: String.t(),
+          vcard_raw: String.t()
+        }
+
   defstruct etag: nil, url: nil, vcard_raw: nil
 
+  @doc """
+  Gets a single contact data.
+  """
+  @spec get(
+          server_url :: String.t(),
+          username :: String.t(),
+          password :: String.t(),
+          addressbook :: String.t(),
+          contanct_url :: String.t()
+        ) :: {:ok, etag :: String.t(), contact :: Contact.t()} | {:error, :not_found} | :error
   def get(server_url, username, password, addressbook, contact_url) do
-    authentication = Excalt.Vcard.Addressbook.build_authentication_header(username, password)
+    authentication = Excalt.Vcard.Helpers.build_authentication_header(username, password)
     req_url = Excalt.Vcard.UrlHelper.build_url(server_url, username, addressbook)
     req_body = Excalt.Vcard.ReqBuilder.get_contact(contact_url)
     req_header = [authentication, {"Depth", "1"}]
     request = Finch.build("REPORT", req_url, req_header, req_body)
 
     case Finch.request(request, ExcaltFinch) do
-      {:ok, %Finch.Response{status: 207, body: body}} ->
-        [contact] = Excalt.Vcard.Parser.parse_contacts_from_addressbook(body)
+      {:ok, %Finch.Response{status: 207, body: body, headers: headers}} ->
+        etag = Excalt.Helpers.Request.extract_from_header(headers, "etag")
+        {:ok, [contact]} = Excalt.Vcard.Parser.parse_contacts_from_addressbook(body)
+        {:ok, etag, contact}
 
       {:ok, %Finch.Response{status: 404, body: _}} ->
         {:error, :not_found}
 
       _ ->
-        # implement more error handling
         :error
     end
   end
 
+  @doc """
+  Creates a new contact, VCF text should be a valid VCARD text
+  If the contact already exists, and we try to do multiple creation of the same contact, the response code 201 is returned but the etag is not changed.
+  """
+  @spec create(
+          server_url :: String.t(),
+          username :: String.t(),
+          password :: String.t(),
+          addressbook_name :: String.t(),
+          vcf_text :: String.t()
+        ) :: {:ok, etag :: String.t()} | {:error, any()}
   def create(server_url, username, password, addressbook_name, vcf_text) do
-    authentication = Excalt.Vcard.Addressbook.build_authentication_header(username, password)
+    authentication = Excalt.Vcard.Helpers.build_authentication_header(username, password)
     req_url = Excalt.Vcard.UrlHelper.build_url(server_url, username, addressbook_name)
     req_body = vcf_text
     contact_uuid = Elixir.UUID.uuid4()
```

Comment:
Ideally, the uuid should be extracted from the vcard directly and only be newly created if it does not exist there.

Reply:
Yes, it will be in the function head to pass, since extraction and actions will be responsibility of the client.